File: src/org/simbrain/network/synapse_update_rules/STDPRule.java
/*
 * Part of Simbrain--a java-based neural network kit
 * Copyright (C) 2005,2007 The Authors.  See http://www.simbrain.net/credits
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package org.simbrain.network.synapse_update_rules;

import org.simbrain.network.core.SpikingNeuronUpdateRule;
import org.simbrain.network.core.Synapse;
import org.simbrain.network.core.SynapseUpdateRule;
import org.simbrain.util.UserParameter;

/**
 * <b>STDPSynapse</b> models spike time dependent plasticity.
 * <p>
 * Only works if source and target neurons are spiking neurons.
 * <p>
 * Drew on: Jean-Philippe Thivierge and Paul Cisek (2008), Journal of
 * Neuroscience. Nonperiodic Synchronization in Heterogeneous Networks of
 * Spiking Neurons. Also drew on the Scholarpedia article.
 */
public class STDPRule extends SynapseUpdateRule {

    // TODO: check description
    /**
     * Time constant for LTD.
     */
    @UserParameter(label = "Tau minus", description = "Time constant " + "for LTD.", increment = .1, order = 0)
    protected double tau_minus = 60;

    /**
     * Time constant for LTP.
     */
    @UserParameter(label = "Tau plus", description = "Time constant " + "for LTP.", increment = .1, order = 1)
    protected double tau_plus = 30;

    /**
     * Learning rate for LTP case. Controls magnitude of LTP changes.
     */
    @UserParameter(label = "W+", description = "Learning rate for " + "LTP case. Controls magnitude of LTP changes.", increment = .1, order = 2)
    protected double W_plus = 10;

    /**
     * Learning rate for LTP case. Controls magnitude of LTD changes.
     */
    @UserParameter(label = "W-", description = "Learning rate for " + "LTP case. Controls magnitude of LTD changes.", increment = .1, order = 3)
    protected double W_minus = 10;

    /**
     * General learning rate.
     */
    @UserParameter(label = "Learning rate", description = "General learning " + "rate.", increment = .1, order = 4)
    protected double learningRate = 0.01;

    /**
     * Sets whether or not STDP acts directly on W or dW/dt
     */
    @UserParameter(label = "Smooth STDP", description = "Whether STDP acts directly on weight or on its derivative instead", order = 5)
    protected boolean continuous = false;

    @Override
    public void init(Synapse synapse) {
    }

    @Override
    public String getName() {
        return "STDP";
    }

    public STDPRule(){

    }

    public STDPRule(STDPRule toCpy) {
        this(toCpy.getW_plus(), toCpy.getW_minus(), toCpy.getTau_plus(),
                toCpy.getTau_minus(), toCpy.learningRate, toCpy.continuous);
    }

    public STDPRule(double w_plus, double w_minus, double tau_plus, double tau_minus, double learningRate, boolean continuous) {
        this.W_plus=w_plus;
        this.W_minus=w_minus;
        this.tau_plus=tau_plus;
        this.tau_minus=tau_minus;
        this.learningRate=learningRate;
        this.continuous=continuous;
    }

    @Override
    public SynapseUpdateRule deepCopy() {
        STDPRule duplicateSynapse = new STDPRule();
        duplicateSynapse.setTau_minus(this.getTau_minus());
        duplicateSynapse.setTau_plus(this.getTau_plus());
        duplicateSynapse.setW_minus(this.getW_minus());
        duplicateSynapse.setW_plus(this.getW_plus());
        duplicateSynapse.setLearningRate(this.getLearningRate());
        duplicateSynapse.setHebbian(hebbian);
        return duplicateSynapse;
    }

    private boolean hebbian = true;

    private double delta_w = 0;

    @Override
    public void update(Synapse synapse) {
        final double str = synapse.getStrength();
        if (synapse.getSource().isSpike() || synapse.getTarget().isSpike()) {
            try {

                final double delta_t = ((((SpikingNeuronUpdateRule) synapse.getSource().getUpdateRule())
                        .getLastSpikeTime())
                        - ((SpikingNeuronUpdateRule) synapse.getTarget().getUpdateRule()).getLastSpikeTime())
                        * (hebbian ? 1 : -1);
                if (delta_t < 0) {
                    delta_w = W_plus * Math.exp(delta_t / tau_plus) * learningRate;
                } else if (delta_t > 0) {
                    delta_w = -W_minus * Math.exp(-delta_t / tau_minus) * learningRate;
                }

            } catch (ClassCastException cce) {
                cce.printStackTrace();
                System.out.println("Don't use non-spiking neurons with STDP!");
            }
            if ( !continuous && Math.signum(str) == -1 ) {
                synapse.setStrength(str - delta_w * synapse.getSource().getNetwork().getTimeStep());
            } else {
                synapse.setStrength(str + delta_w * synapse.getSource().getNetwork().getTimeStep());
            }
        }

        if ( continuous && Math.signum(str) == -1 ) {
            synapse.setStrength(str - delta_w * synapse.getSource().getNetwork().getTimeStep());
        } else {
            synapse.setStrength(str + delta_w * synapse.getSource().getNetwork().getTimeStep());
        }
    }

    public double getTau_plus() {
        return tau_plus;
    }

    public void setTau_plus(double tauPlus) {
        tau_plus = tauPlus;
    }

    public double getTau_minus() {
        return tau_minus;
    }

    public void setTau_minus(double tauMinus) {
        tau_minus = tauMinus;
    }

    public double getW_plus() {
        return W_plus;
    }

    public void setW_plus(double wPlus) {
        W_plus = wPlus;
    }

    public double getW_minus() {
        return W_minus;
    }

    public void setW_minus(double wMinus) {
        W_minus = wMinus;
    }

    public double getLearningRate() {
        return learningRate;
    }

    public void setLearningRate(double learningRate) {
        this.learningRate = learningRate;
    }

    public boolean isHebbian() {
        return hebbian;
    }

    public void setHebbian(boolean hebbian) {
        this.hebbian = hebbian;
    }

    public boolean isContinuous() {
        return continuous;
    }

    public void setContinuous(boolean continuous) {
        this.continuous=continuous;
    }

    public double getDelta_w() {
        return  delta_w;
    }

    public void setDelta_w(double delta_w) {
        this.delta_w = delta_w;
    }

}

File: src/main/kotlin/org/simbrain/network/learningrules/STDPRule.kt
/*
 * Part of Simbrain--a java-based neural network kit
 * Copyright (C) 2005,2007 The Authors.  See http://www.simbrain.net/credits
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package org.simbrain.network.learningrules

import org.simbrain.network.core.*
import org.simbrain.network.util.EmptyMatrixData
import org.simbrain.network.util.EmptyScalarData
import org.simbrain.network.util.SpikingMatrixData
import org.simbrain.util.UserParameter
import kotlin.math.exp

/**
 * Models spike time dependent plasticity STDP.
 *
 * Assumes source and target neurons are spiking neurons.
 *
 * See [StdpSim.kt] for more information about the rule.
 *
 * Sources: Jean-Philippe Thivierge and Paul Cisek (2008), Nonperiodic Synchronization in Heterogeneous Networks of  Spiking Neurons
 * and the Scholarpedia article on STDP.
 *
 * Also on anti-hebbian stdp: https://journals.physiology.org/doi/pdf/10.1152/jn.00551.2006
 */
open class STDPRule : SynapseUpdateRule<EmptyScalarData, EmptyMatrixData> {

    @UserParameter(
        label = "Tau plus",
        description = "Time constant for LTP (weight strengthening when pre fires before post. Smaller values narrow the window within which LTP is applied.",
        increment = .1,
        order = 10)
    var tauPlus: Double = 30.0

    // Often wider than tau plus in the literature
    @UserParameter(
        label = "Tau minus",
        description = "Time constant for LTD (weight decay when post fires before pre). Smaller values narrow the window within which LTD is applied.",
        increment = .1,
        minimumValue = 0.0,
        order = 20)
    var tauMinus: Double = 60.0

    @UserParameter(
        label = "W+",
        description = "Learning rate for LTP case. Controls the magnitude of LTP changes.",
        increment = .1,
        minimumValue = 0.0,
        order = 30
    )
    open var wPlus: Double = 10.0

    @UserParameter(
        label = "W-",
        description = "Learning rate for LTD. Controls magnitude of LTD changes",
        increment = .1,
        minimumValue = 0.0,
        order = 40
    )
    open var wMinus: Double = 10.0

    @UserParameter(
        label = "Learning rate",
        description = "Global learning rate",
        increment = .1,
        minimumValue = 0.0,
        order = 50)
    var learningRate: Double = 0.01

    @UserParameter(
        label = "Hebbian",
        description = "If true, use hebbian learning, else anti-hebbian",
        order = 60
    )
    var isHebbian: Boolean = true

    override fun init(synapse: Synapse) {
    }

    override val name: String
        get() = "STDP"

    constructor()

    constructor(
        w_plus: Double,
        w_minus: Double,
        tau_plus: Double,
        tau_minus: Double,
        learningRate: Double,
        continuous: Boolean
    ) {
        this.wPlus = w_plus
        this.wMinus = w_minus
        this.tauPlus = tau_plus
        this.tauMinus = tau_minus
        this.learningRate = learningRate
    }

    override fun copy(): SynapseUpdateRule<*, *> {
        val duplicateSynapse = STDPRule()
        duplicateSynapse.tauMinus = tauMinus
        duplicateSynapse.tauPlus = tauPlus
        duplicateSynapse.wMinus = wMinus
        duplicateSynapse.wPlus = wPlus
        duplicateSynapse.learningRate = learningRate
        duplicateSynapse.isHebbian = isHebbian
        return duplicateSynapse
    }

    open var deltaW: Double = 0.0

    context(Network)
    override fun apply(synapse: Synapse, data: EmptyScalarData) {
        if (synapse.source.isSpike || synapse.target.isSpike) {
            val deltaT = ((synapse.source.lastSpikeTime
                    - synapse.target.lastSpikeTime)
                    * (if (isHebbian) 1 else -1))
            if (deltaT < 0) {
                // LTP Case
                deltaW = wPlus * exp(deltaT / tauPlus) * learningRate
            } else if (deltaT > 0) {
                // LTD Case
                deltaW = -wMinus * exp(-deltaT / tauMinus) * learningRate
            }

            synapse.strength +=  deltaW * timeStep
        }

    }

    context(Network)
    override fun apply(connector: Connector, dataHolder: EmptyMatrixData) {
        val weightMatrix = connector as? WeightMatrix ?: return
        val sourceNeuronArray = weightMatrix.source as? NeuronArray ?: return
        val targetNeuronArray = weightMatrix.target as? NeuronArray ?: return
        
        // Ensure both neuron arrays have spiking data
        val sourceSpikingData = sourceNeuronArray.dataHolder as? SpikingMatrixData ?: return
        val targetSpikingData = targetNeuronArray.dataHolder as? SpikingMatrixData ?: return
        
        // Only apply if either source or target neurons have spiked
        val hasSourceSpikes = sourceSpikingData.spikes.any { it }
        val hasTargetSpikes = targetSpikingData.spikes.any { it }
        
        if (hasSourceSpikes || hasTargetSpikes) {
            // Get spike times
            val sourceSpikeTimes = sourceSpikingData.lastSpikeTimes
            val targetSpikeTimes = targetSpikingData.lastSpikeTimes
            
            // Apply STDP rule to each connection
            for (i in 0 until weightMatrix.weights.nrow()) { // target neurons
                for (j in 0 until weightMatrix.weights.ncol()) { // source neurons
                    // Only update if at least one neuron has spiked
                    if (sourceSpikingData.spikes[j] || targetSpikingData.spikes[i]) {
                        val deltaT = ((sourceSpikeTimes[j] - targetSpikeTimes[i]) 
                                * (if (isHebbian) 1 else -1))
                        
                        val deltaW = if (deltaT < 0) {
                            // LTP Case
                            wPlus * exp(deltaT / tauPlus) * learningRate
                        } else if (deltaT > 0) {
                            // LTD Case
                            -wMinus * exp(-deltaT / tauMinus) * learningRate
                        } else {
                            0.0
                        }
                        
                        weightMatrix.weights[i, j] += deltaW * timeStep
                    }
                }
            }
        }
    }
}
